Complete system simulation using PyBullet

gizatt opened the pull request:
A monster PR! Not review-ready yet, I'll probably need to find a way to split out the key features, since some of my dish simulation stuff got mixed in here.

Gist of what is added:
- `drake_simulation` package provides Pybullet and Drake simulations from YAML simulation-context files. These list out object instances and robot parameters to be simulated, and like in `drake_simulation/config.` This includes entirely passive simulations in both engines (to compare their performance), as well as a partial system sim in Drake (everything but point clouds), and a complete system sim in Pybullet.
- Procman includes a sim startup script that results in an environment that should look identical to the robot, except it's running at ~20% real time (depending on rendering settings... physics simulation is > real time but 640x480 rendering is mysteriously slow).
- Added a python script for launching a complete robot context (i.e. a procman deputy, a system with sim + controllers) and running basic movement and IK commands through their convenient Python bindings in `modules/spartan`. Enabled on Jenkins starting with this PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/spartan/240)
<!-- Reviewable:end -->

gizatt commented:
Jenkins please test

gizatt commented:
Jenkins please test

gizatt commented:
yessss

gizatt commented:
This is now ready for review, if anyone is brave enough. I would be happy to try to chunk this to make it more tractable to understand its pieces, but it's been sitting for long enough that I definitely wouldn't mind getting it in master and resolving difficulties as they arise.

peteflorence commented:
This looks awesome Greg I can gladly review tonight

On Wed, Jan 31, 2018 at 5:41 PM Gregory Izatt <notifications@github.com>
wrote:

> This is now ready for review, if anyone is brave enough. I would be happy
> to try to chunk this to make it more tractable to understand its pieces,
> but it's been sitting for long enough that I definitely wouldn't mind
> getting it in master and resolving difficulties as they arise.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RobotLocomotion/spartan/pull/240#issuecomment-362095804>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFYQqBrySa5Epf1qHX1wzkdx2Jz8wK1xks5tQOwXgaJpZM4RsH9f>
> .
>

peteflorence commented on CMakeLists.txt:
```diff
@@ -26,13 +26,17 @@ option(WITH_SNOPT "Build drake with support for the SNOPT solver." ON)
 option(WITH_MOSEK "Build with support for MOSEK" OFF)
 
+# Intended to install the SAME version as Drake uses internally.
+# However, the SHAs are currently not synced (I don't know how to
```
@gizatt what are the implications, if you've found any, of not having synced bullet versions with drake?

gizatt replied:
Haven't found any. The things it would hurt would be those few things that consume the outside-defined bullet: which at the moment, is just the Pybullet simulation stuff. It could introduce linker errors in those applications if it gets really out of step, but I think you'd have to write a (feasible but) Frankenstein C++ application for that to be likely?

peteflorence commented on apps/iiwa/kuka_iiwa_state_translator.py:
```diff
@@ -11,8 +11,9 @@
 class KukaIiwaStateTranslator(object):
 
-    def __init__(self):
+    def __init__(self, basePosition):
```
@gizatt are we sure any users of KukaIiwaStateTranslator are now using this constructor with a `basePosition`?

gizatt replied:
Sim does! I'm using the same configs that Drake uses in their sims, which place the world origin on the ground under the IIWA table, rather than at IIWA origin. This is intended to be transparent for the rest of the system (kuka_iiwa_state_translator has a default arg for this).

peteflorence commented on models/dishes/bowl_6p25in.urdf:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0"?>
```
Do we think we want all of these objects git-version-controlled?

Would it be better to have them .gitignored, and `wget`ted via a `./get_dish_mesges.sh`?

They aren't too too big / not sure how often they'll change

peteflorence replied:
@gizatt -- after seeing that the dish rack is over 30,000 lines... yes I'd vote to have all of `models/dishes` be downloaded outside of git.  What do you think?  Or at least the big ones?

peteflorence commented on modules/spartan/manipulation/director_schunk_driver.py:
```diff
@@ -19,4 +19,4 @@ def sendCloseGripperCommand(self):
 	def sendGripperCommand(self, position, force):
-		self.taskRunner.callOnThread(self.schunkDriver.sendGripperCommand, posiition, force)
\ No newline at end of file
+		self.taskRunner.callOnThread(self.schunkDriver.sendGripperCommand, position, force)
```
good catch :) `posiition`

manuelli replied:
yeah that must have been my bad. Thanks for the catch

peteflorence commented on setup/docker/test_full_simulation_stack.py:
```diff
@@ -0,0 +1,143 @@
+        cleanup_and_exit(1)
+
+
+def try_to_move_arm_and_hand():
```
love this

peteflorence commented on src/catkin_projects/drake_simulation/CMakeLists.txt:
```diff
@@ -0,0 +1,165 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(drake_simulation)
```
what is the `drake_simulation` package's job meant to be?

gizatt replied:
Yeah, I ought to rename this before it comes in... originally it was just a full stack simulation using Drake, but now it's the home for most of the simulation tools (as you saw below)

peteflorence commented on src/catkin_projects/drake_simulation/README.md:
```diff
@@ -0,0 +1,55 @@
+Simulation Tools
+=================
+Provides tools for simulation in Spartan.
```
ah.  nice.

peteflorence commented on src/catkin_projects/robot_control/src/robot_control/robotmovementservice.py:
```diff
@@ -47,7 +47,15 @@ def setupRobot(self):
     def getRobotState(self):
-        return self.subscribers['joint_states'].waitForNextMessage()
+        full_state = self.subscribers['joint_states'].waitForNextMessage()
+        # Reduce to just this robot's state
```
@gizatt what else was using getRobotState?

gizatt replied:
From a search across Spartan: it *looks* like no-one, but this is a change Lucas will be very interested in.

The underlying issue was that I've made previous changes in Spartan to have the state publisher glue together the Schunk and IIWA states into one large robot state. However, the movement service doesn't actually care about the Schunk state (and its joint goals usually only specify IIWA joint goals anyway)... which winds up causing size mismatches between state vectors in the movement service movement routines.

manuelli commented:
I'll try to give this a thorough look today. Sorry for the delay

gizatt commented:
- Moved both rlg_misc_models and dish_models to git submodules, hosted (for the moment?) on my personal github.
- Rename drake_simulation to rlg_simulation, and add a CMake switch to build the pure-Drake IIWA sim that I wound up only using for comparison purposes. (It comprises many files, is likely to break, and is likely to become redundant as Drake gets better -- I want it around for reference, but not under CI or usually built, so that switch defaults to off.)

manuelli commented on modules/spartan/manipulation/schunk_driver.py:
```diff
@@ -41,6 +40,11 @@ def setupDefaultMessages(self):
+	def handleStatusCallback(self, msg):
+		self.lastStatusMsg = msg
+		if self.statusSubscriberCallback is not None:
+			self.statusSubscriberCallback(msg)
```
ok this is cleaner than what I had before. Looks good

manuelli commented:
Reviewed 1 of 148 files at r7.
Review status: 1 of 59 files reviewed at latest revision, 12 unresolved discussions.

---

*[CMakeLists.txt, line 151 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gEUJPG9kqnMiQXNeF:-L4gEUJPG9kqnMiQXNeG:b-1jnko1) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/CMakeLists.txt#L151)):*
> ```CMake
>     -DPYBIND11_TEST:BOOL=OFF
> )
> ```

Why do we need pybind? Is it so that you can use pybullet?

---

*[CMakeLists.txt, line 228 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gEeLYOTv6zWhUi8G9:-L4gEeLYOTv6zWhUi8GA:b-paaly4) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/CMakeLists.txt#L228)):*
> ```CMake
>   # install after Drake. We're installing a different
>   # version than Drake (2.86.1 vs post-2.87), which is risky
>   # enough already.
> ```

do you think they will have runtime conflicts? e.g. building against drake and then finding this one at runtime?

---

*[CMakeLists.txt, line 357 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gFq7ivHkFOtpatmrA:-L4gFq7jVw7Q2lDicSer:b88ftpq) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/CMakeLists.txt#L357)):*
> ```CMake
>       # Combine environment prefix path (with ROS) with our own
>       "-DCMAKE_PREFIX_PATH:PATH=${CMAKE_INSTALL_PREFIX};${CMAKE_PREFIX_PATH_ENV}"
>       -DWITH_DRAKE_IIWA_SIM:BOOL=${WITH_DRAKE_IIWA_SIM}
> ```

is this an arg that's passed into the catkin-build? We should really switch to using catkin_tools, this way of building ROS packages is pretty wonky.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 2 of 148 files at r7.
Review status: 3 of 59 files reviewed at latest revision, 13 unresolved discussions.

---

*[apps/iiwa/iiwa_hardware.pmd, line 12 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gG2_ltgIlKOxi8rA9:-L4gG2_ltgIlKOxi8rAA:bjwmc7z) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/apps/iiwa/iiwa_hardware.pmd#L12)):*
> ```
>   }
>   
>   cmd "1.pybullet_kuka_simulation_single_object_ci" {
> ```

you have two things named cmd 1.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 1 of 42 files at r5.
Review status: 4 of 59 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 1 of 42 files at r5, 1 of 148 files at r7.
Review status: 6 of 59 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 1 of 42 files at r5, 1 of 148 files at r7.
Review status: 8 of 59 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 1 of 139 files at r1, 2 of 42 files at r5.
Review status: 11 of 59 files reviewed at latest revision, 15 unresolved discussions.

---

*[modules/spartan/utils/ros_utils.py, line 170 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gH7VmlNGTffW7hJV7:-L4gH7VmlNGTffW7hJV8:b-90vxbu) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/modules/spartan/utils/ros_utils.py#L170)):*
> ```Python
>         self.numJoints = len(jointNames)
> 
>     def moveToJointPosition(self, q, maxJointDegreesPerSecond=30, timeout=None):
> ```

I don't think None is a good default argument here. This could lead to dangerous behavior if you have a stale call of this on a thread and later the ROS service comes back online. Can we set this to something reasonable like 10 seconds?

---

*[modules/spartan/utils/ros_utils.py, line 181 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gHJIt_grzjCULE07s:-L4gHJIt_grzjCULE07t:b-ia41wb) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/modules/spartan/utils/ros_utils.py#L181)):*
> ```Python
>         return response
> 
>     def moveToCartesianPosition(self, poseStamped, maxJointDegreesPerSecond=30, timeout=None):
> ```

same comment as above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Review status: 11 of 59 files reviewed at latest revision, 16 unresolved discussions.

---

*[modules/spartan/utils/ros_utils.py, line 198 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gHMALnfeL5GU5BWHx:-L4gHMAM_xB-l7NPPuio:b-ia41wb) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/modules/spartan/utils/ros_utils.py#L198)):*
> ```Python
>         return self.moveToJointPosition(joint_state.position, maxJointDegreesPerSecond=maxJointDegreesPerSecond)
> 
>     def runIK(self, poseStamped, seedPose=None, nominalPose=None, timeout=None):
> ```

same comment as above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 1 of 42 files at r5.
Review status: 12 of 59 files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 1 of 139 files at r1.
Review status: 13 of 59 files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Review status: 13 of 59 files reviewed at latest revision, 17 unresolved discussions.

---

*[setup/docker/run_jenkins.sh, line 43 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gHZnoRm6oLW_VeRTr:-L4gHZnoRm6oLW_VeRTs:bmyc6zk) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/setup/docker/run_jenkins.sh#L43)):*
> ```Shell
> 
> # Launch a complete robot context and execute some canned movement.
> DISPLAY=:100 python ${SPARTAN_SOURCE_DIR}/setup/docker/test_full_simulation_stack.py
> ```

nice, having an end to end test is great.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 1 of 42 files at r5.
Review status: 14 of 59 files reviewed at latest revision, 17 unresolved discussions.

---

*[setup/docker/run_jenkins.sh, line 43 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gHZnoRm6oLW_VeRTr:-L4gHi5hFdggO6bUue-u:bb6ywmx) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/setup/docker/run_jenkins.sh#L43)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

nice, having an end to end test is great.
</blockquote></details>

LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 1 of 42 files at r5.
Review status: 15 of 59 files reviewed at latest revision, 17 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Review status: 15 of 59 files reviewed at latest revision, 18 unresolved discussions.

---

*[src/catkin_projects/drake_ros_systems, line 1 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gIMwjIFFmHPJ7BfOo:-L4gIMwjIFFmHPJ7BfOp:b-8hnlra) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/src/catkin_projects/drake_ros_systems#L1)):*
> ```
> Subproject commit b2a32ef1d88ec8ff72a4d17d30850d074549b0c3
> ```

what is this exactly? Just publishers/subscribers for LCM/ROS? We should get permissions from Russ to put everything under RobotLocomotion organization, or something else owned by lab and not on personal githubs I think. But we can punt on this for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 3 of 139 files at r1, 1 of 3 files at r3, 1 of 42 files at r5, 2 of 148 files at r7.
Review status: 21 of 59 files reviewed at latest revision, 18 unresolved discussions.

---

*[src/catkin_projects/robot_control/src/robot_control/robotmovementservice.py, line 51 at r6](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L3ebEPAsUyhAdhw65m--r6-51:-L4gJcqafmiNcL8jovQq:batjuve) ([raw file](https://github.com/robotlocomotion/spartan/blob/bceeb4700eeb51d69696c7b2a1a6467170056a9b/src/catkin_projects/robot_control/src/robot_control/robotmovementservice.py#L51)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

From a search across Spartan: it *looks* like no-one, but this is a change Lucas will be very interested in.

The underlying issue was that I've made previous changes in Spartan to have the state publisher glue together the Schunk and IIWA states into one large robot state. However, the movement service doesn't actually care about the Schunk state (and its joint goals usually only specify IIWA joint goals anyway)... which winds up causing size mismatches between state vectors in the movement service movement routines.
</blockquote></details>

Ok I see what you did here. I would vote to add one more safety however. If anything in self.jointNames isn't in full_state.name then we have a problem. Basically reduced_state should always have size len(self.jointNames).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 1 of 42 files at r5.
Review status: 22 of 59 files reviewed at latest revision, 19 unresolved discussions.

---

*[src/catkin_projects/robot_server/src/robot_server/pub_joint_state.py, line 76 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gK_GDPjM4DhPuEKVc:-L4gK_GDPjM4DhPuEKVd:babvsi3) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/src/catkin_projects/robot_server/src/robot_server/pub_joint_state.py#L76)):*
> ```Python
>         # messing up the TF server), so for now I'm approximating with
>         # our local system time. -gizatt
>         self.publishROSJointStateMessage()
> ```

looks like publishROSJointStateMessage() never took an arg, so why was this working before . . . 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 1 of 42 files at r5.
Review status: 23 of 59 files reviewed at latest revision, 19 unresolved discussions.

---

*[models/dishes/bowl_6p25in.urdf, line 1 at r6](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L3ebEP6KxmSbsAl169g-r6-1:-L4gKjZJ5DmkBh_f5cYh:bgj5a6i) ([raw file](https://github.com/robotlocomotion/spartan/blob/bceeb4700eeb51d69696c7b2a1a6467170056a9b/models/dishes/bowl_6p25in.urdf#L1)):*
<details><summary><i>Previously, peteflorence (Pete Florence) wrote…</i></summary><blockquote>

@gizatt -- after seeing that the dish rack is over 30,000 lines... yes I'd vote to have all of `models/dishes` be downloaded outside of git.  What do you think?  Or at least the big ones?
</blockquote></details>

can this be moved into the dish_models repo?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Ok just did a first pass. I reviewed the critical pieces, didn't look at the details of the simulator stuff. A few minor comments but overall ok.

---

Reviewed 1 of 148 files at r7.
Review status: 23 of 59 files reviewed at latest revision, 19 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-:-L4gKsGoO86L5ptY7EZn:b-e7xhn0)*
<!-- Sent from Reviewable.io -->

gizatt commented:
Review status: 20 of 59 files reviewed at latest revision, 19 unresolved discussions.

---

*[CMakeLists.txt, line 151 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gEUJPG9kqnMiQXNeF:-L4gy-QekhrXGVxhnFez:b11r5l0) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/CMakeLists.txt#L151)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

Why do we need pybind? Is it so that you can use pybullet?
</blockquote></details>

I've added python bindings for the RemoteTreeViewer, since we have an increasing amount of pydrake floating around

---

*[CMakeLists.txt, line 228 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gEeLYOTv6zWhUi8G9:-L4gy7ysJeq2m_ItlBXe:bm22bwh) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/CMakeLists.txt#L228)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

do you think they will have runtime conflicts? e.g. building against drake and then finding this one at runtime?
</blockquote></details>

I haven't encountered any so far in testing different parts of the system, including large perception processes that link Drake.

---

*[CMakeLists.txt, line 357 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gFq7ivHkFOtpatmrA:-L4gyu6h5yutCM18xFcS:brmnfk6) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/CMakeLists.txt#L357)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

is this an arg that's passed into the catkin-build? We should really switch to using catkin_tools, this way of building ROS packages is pretty wonky.
</blockquote></details>

Agreed, this is cludgy, but only way I have to pass options right now.

---

*[apps/iiwa/iiwa_hardware.pmd, line 12 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gG2_ltgIlKOxi8rA9:-L4gz0rfKUbUye60VvHC:b-w44zqw) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/apps/iiwa/iiwa_hardware.pmd#L12)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

you have two things named cmd 1.
</blockquote></details>

Fixed, but numbering these is pretty fragile (as every time anyone wants to add something not at the end of the process list, a bunch of names need to get changed, which is really likely to break a script).

---

*[modules/spartan/utils/ros_utils.py, line 170 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gH7VmlNGTffW7hJV7:-L4gzlFzAwj9iP4Qseqb:b-896fix) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/modules/spartan/utils/ros_utils.py#L170)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

I don't think None is a good default argument here. This could lead to dangerous behavior if you have a stale call of this on a thread and later the ROS service comes back online. Can we set this to something reasonable like 10 seconds?
</blockquote></details>

Done.

---

*[modules/spartan/utils/ros_utils.py, line 181 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gHJIt_grzjCULE07s:-L4gzlougaJFwRLqeIwu:b-896fix) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/modules/spartan/utils/ros_utils.py#L181)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

same comment as above
</blockquote></details>

Done.

---

*[modules/spartan/utils/ros_utils.py, line 198 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gHMALnfeL5GU5BWHx:-L4gzm9i5sNUCdk_cPmd:b-896fix) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/modules/spartan/utils/ros_utils.py#L198)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

same comment as above
</blockquote></details>

Done.

---

*[src/catkin_projects/drake_ros_systems, line 1 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gIMwjIFFmHPJ7BfOo:-L4gztgqWDgzMA3NtFwN:bk538k1) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/src/catkin_projects/drake_ros_systems#L1)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

what is this exactly? Just publishers/subscribers for LCM/ROS? We should get permissions from Russ to put everything under RobotLocomotion organization, or something else owned by lab and not on personal githubs I think. But we can punt on this for now.
</blockquote></details>

Right, this in particular is a collection of drake/ROS interoperation that I think TRI (particularly Naveen?) has their own version of in the works.

---

*[src/catkin_projects/rlg_simulation/CMakeLists.txt, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L3ebEP7iJh4a7_o12jq-r6-2:-L4h--ej-ZZ7sqvNqOq1:ba3o3s1) ([raw file](https://github.com/robotlocomotion/spartan/blob/bceeb4700eeb51d69696c7b2a1a6467170056a9b/src/catkin_projects/rlg_simulation/CMakeLists.txt#L2)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Yeah, I ought to rename this before it comes in... originally it was just a full stack simulation using Drake, but now it's the home for most of the simulation tools (as you saw below)
</blockquote></details>

Renamed

---

*[src/catkin_projects/robot_control/src/robot_control/robotmovementservice.py, line 51 at r6](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L3ebEPAsUyhAdhw65m--r6-51:-L4h1r03xgUzSrIr9JQl:b-896fix) ([raw file](https://github.com/robotlocomotion/spartan/blob/bceeb4700eeb51d69696c7b2a1a6467170056a9b/src/catkin_projects/robot_control/src/robot_control/robotmovementservice.py#L51)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

Ok I see what you did here. I would vote to add one more safety however. If anything in self.jointNames isn't in full_state.name then we have a problem. Basically reduced_state should always have size len(self.jointNames).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Review status: 20 of 59 files reviewed at latest revision, 17 unresolved discussions.

---

*[CMakeLists.txt, line 357 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gFq7ivHkFOtpatmrA:-L4h9VmEYKDrUoZZMtfA:b-3xl9xf) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/CMakeLists.txt#L357)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Agreed, this is cludgy, but only way I have to pass options right now.
</blockquote></details>

ok lets make an issue to track this, I'm ok with it as a temporary fix. I would like to investigate switching to catkin_tools eventually. If John Carter returns to the lab that could be a good project for him.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Review status: 20 of 59 files reviewed at latest revision, 16 unresolved discussions.

---

*[apps/iiwa/iiwa_hardware.pmd, line 12 at r7](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L4gG2_ltgIlKOxi8rA9:-L4h9gViBwmEe3hS0wP2:bdmm2sb) ([raw file](https://github.com/robotlocomotion/spartan/blob/f088b21ab92ae36a5c30c4cbe04586857f7905a9/apps/iiwa/iiwa_hardware.pmd#L12)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Fixed, but numbering these is pretty fragile (as every time anyone wants to add something not at the end of the process list, a bunch of names need to get changed, which is really likely to break a script).
</blockquote></details>

yes, it's very annoying. Ok for now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Review status: 20 of 59 files reviewed at latest revision, 13 unresolved discussions.

---

*[setup/docker/test_full_simulation_stack.py, line 46 at r6](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-L3ebEPAsUyhAdhw65mB-r6-46:-L4h9oiU1msJuqGr1f4y:b-k3pv94) ([raw file](https://github.com/robotlocomotion/spartan/blob/bceeb4700eeb51d69696c7b2a1a6467170056a9b/setup/docker/test_full_simulation_stack.py#L46)):*
<details><summary><i>Previously, peteflorence (Pete Florence) wrote…</i></summary><blockquote>

love this
</blockquote></details>

nice :+1: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
Reviewed 2 of 148 files at r7.
Review status: 21 of 59 files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240)*
<!-- Sent from Reviewable.io -->

manuelli commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 21 of 59 files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/240#-:-L4hA9Ma59y9iqcxBV1a:bnfp4nl)*
<!-- Sent from Reviewable.io -->
